doc/feature/compartmentReport.md

```diff
@@ -0,0 +1,467 @@
+        * - There is no previous read or seek operation with a pending future.
+        *
+        * Postconditions:
+        * Let:
+        *  - r be the CompartmentReport
+        *  - f be the returned future by r.readNextFrame()
+        *  - and s = r.getCurrentTime() before read is called (s = -inf if
+        *    undefined):
+        * After f.wait() returns the following are all true:
+        *  - If r.getState() == ENDED s == r.getCurrentTime() and f.get() is
+        *    empty
+        *  - Otherwise r.getCurrentTime() ~= s + r.getTimestep() (the timestamp
+        *    is rounded to the nearest multiple of the timesetp).
+        *  - The mapping data becomes defined.
+        *
+        * After successful f.wait_for or f.wait_until, the result is the same
+        * as above. If they time out, no observable state changes.
+        *
+        * @throw std::logic_error if a precondition is not met.
+        * @version TBD
+        */
+       boost::future< floatsPtr > readNextFrame( );
+
+- Current write functions have inconsistent return types: writeHeader is void
+  whereas writeCompartments, writeFrame and flush return bool. The proposal
+  is to make all new write functions void and use exceptions to report write
+  errors instead. This is consistent with read functions using exceptions to
+  report errors.
+
+- Remove the timestamp argument from writeFrame to enforce sequential writing.
+  If random access is needed for any reason client code could still use seek
+  for that purpose. The motivation for this change is the simplification of
+  the stream implementation while making it easier to detect unsupported
+  operations (e.g. seek backwards in a stream). To avoid confusion, writeFrame
+  is renamed to writeCell. The new function is:
+
+       /** Write the voltage for once cell at a point in time.
+        *
+        * Preconditions:
+        * - The report was open in write mode.
+        * - writeCellMapping has already been called.
+        *
+        * @param gid the GID of the cell
+        * @param voltages the voltages per compartment to save
+        * @throw std::logic_error if a precondition is not met and
+        *        std::runtime_eror if an IO error occurs.
+        * @version TBD
+        */
+       void writeCell( const uint32_t gid, const floats& voltages );
+
+  Replace writeCompartments with writeCellMapping, with a new return type
+  and additional preconditions:
+
+       /** Write the compartment counts for each section for one cell.
+        *
+        * Preconditions:
+        * - The report was open in write mode.
+        * - writeHeader has already been called.
+        *
+        * @param gid the GID of the cell
+        * @param counts the number of compartments per section
+        * @throw std::logic_error if a precondition is not met and
+        *        std::runtime_eror if an IO error occurs.
+        * @version TBD
+        */
+       void writeCellMapping( const uint32_t gid,
```

Comment:
It updates the internal mapping, but with this call only it's more like writeCellComparments() or writeCellCompartmentCounts(). I'm not sure the term 'mapping' is helpful for a user with our backgrounds.